cmd/fips/main.go

```diff
@@ -0,0 +1,237 @@
+
+	go run ./cmd/fips crypto/...
+
+The GOOS can also be specified. This allows, for example, examining the openssl APIs while running Windows:
+
+	go run ./cmd/fips crypto/... -goos linux
+`
+
+type fnReport struct {
+	PackageID        string
+	Name             string
+	HasBoringEnabled bool
+	BoringCalls      []string
+}
+
+var goos = flag.String("goos", "", "The operating system for which to compile the examined packages. Defaults to GOOS")
+
+func main() {
+	log.SetFlags(0)
+	log.SetPrefix("fips: ")
+
+	flag.Usage = func() {
+		fmt.Fprintf(flag.CommandLine.Output(), "\nUsage:\n")
+		flag.PrintDefaults()
+		fmt.Fprintf(flag.CommandLine.Output(), "%s\n\n", description)
+	}
+
+	flag.Parse()
+
+	pkgs, err := parsePackages()
+	if err != nil {
+		log.Fatalf("load: %v\n", err)
+		os.Exit(1)
+	}
+	if packages.PrintErrors(pkgs) > 0 {
+		os.Exit(1)
+	}
+
+	for _, pkg := range pkgs {
+		// We are only interested in public packages.
+		if strings.Contains(pkg.ID, "/internal/") {
+			continue
+		}
+		tryResolveMissingObjects(pkg.Syntax)
+		for _, src := range pkg.Syntax {
+			for _, decl := range src.Decls {
+				if decl, ok := decl.(*ast.FuncDecl); ok {
+					if decl.Body == nil {
+						continue
+					}
+					if decl.Recv != nil {
+						// We are only interested in top level functions.
+						// Methods may contain boring calls but these are always
+						// called from top level functions.
+						continue
+					}
+					if !ast.IsExported(decl.Name.Name) {
+						// We are only interested in exported declarations.
+						continue
+					}
+					report := fnReport{
+						PackageID: pkg.ID,
+						Name:      decl.Name.Name,
+					}
+					processFuncDecl(decl, &report)
+					log.Println(report)
+				}
+			}
+		}
+	}
+}
+
+func parsePackages() ([]*packages.Package, error) {
+	env := os.Environ()
+	if *goos != "" {
+		env = append(env, "GOOS="+*goos)
+	}
+	cfg := &packages.Config{
+		Env: env,
+		Mode: packages.NeedImports | packages.NeedDeps | packages.NeedSyntax |
+			packages.NeedName | packages.NeedFiles | packages.NeedCompiledGoFiles |
+			packages.NeedTypes | packages.NeedTypesInfo | packages.NeedTypesSizes,
+	}
+	pkgs, err := packages.Load(cfg, flag.Args()...)
+	if err != nil {
+		return nil, err
+	}
+	return pkgs, nil
+}
+
+// Ideally objects are always resolved by the package loader but that's not always true
+// so we try to resolve objects looking at other files within the same package.
+func tryResolveMissingObjects(files []*ast.File) {
+	funcDecls := make(map[string]*ast.Object)
+	for _, f := range files {
+		for _, d := range f.Decls {
+			if fn, ok := d.(*ast.FuncDecl); ok {
+				if fn.Recv == nil && fn.Name.Obj != nil {
+					funcDecls[fn.Name.Name] = fn.Name.Obj
+				}
+			}
+		}
+	}
+	for _, f := range files {
+		ast.Inspect(f, func(n ast.Node) bool {
+			switch n := n.(type) {
+			case *ast.CallExpr:
+				if ident, ok := n.Fun.(*ast.Ident); ok && ident.Obj == nil {
+					if ident.Obj, ok = funcDecls[ident.Name]; ok {
+						return false
```

Comment:
I'm a little lost trying to find what the function is actually modifying. What I'm expecting to see somewhere is a statement something like `funcDecls[ident.Name].SomethingRef = somethingWeFound`, or some other modification to existing nodes that forces them to get resolved.

Is simply traversing the tree "automatically" causing objects to get resolved, and that's all we need to do here? (Should be noted in func signature if so, IMO.)

Reply:
`funcDecls` is just a cache of function declarations, so unresolved indent objects can be resolved later on by just querying the cache using the function name.

The real assignation happens here: `ident.Obj, ok = funcDecls[ident.Name]`. `ident` is a pointer to a node whose function object is undefined, but we have its name, so we can try to resolve it using the declarations in `funcDecls`.

I can see where the confusion come from, so I've added some comments and simplified the code.